apollo-federation/src/type_merging.rs

```diff
@@ -0,0 +1,1364 @@
+                    if is_input_position {
+                        // input: use more general (supertype) → upgrade to source_type
+                        result_type = Some(source_type.clone());
+                    }
+                    // else output: keep current_type (the subtype)
+                } else if let Ok(true) = self.is_strict_subtype(source_type, current_type) {
+                    // source_type is a subtype of current_type (current_type is more general)
+                    has_subtypes = true;
+                    if !is_input_position {
+                        // For input types, use the more specific type (contravariance)
+                        result_type = Some(source_type.clone());
+                    }
+                    // else output: keep current_type (the supertype)
+                } else {
+                    // Types are incompatible
+                    has_incompatible = true;
+                }
+            } else {
+                // First type we encounter
+                result_type = Some(source_type.clone());
+            }
+        }
+
+        let Some(typ) = result_type else {
+            // No sources provided - this shouldn't happen in normal operation
+            return Err(FederationError::SingleFederationError(
+                SingleFederationError::Internal {
+                    message: format!(
+                        "No source provided for {} across subgraphs",
+                        dest.coordinate()
+                    ),
+                },
+            ));
+        };
+
+        // Copy the type reference to the destination schema
+        let copied_type = match self.copy_type_reference(&typ, self.schema) {
+            Ok(copied) => copied,
+            Err(_) => {
+                // If copying fails, use the original type
+                typ.clone()
+            }
+        };
+
+        dest.set_type(copied_type);
+
+        self.track_enum_usage(&typ, dest.coordinate(), is_input_position, sources);
+
+        let element_kind = if is_input_position {
+            "argument"
+        } else {
+            "field"
+        };
+
+        if has_incompatible {
+            // Report incompatible type error
+            let error_code_str = if is_input_position {
+                "ARGUMENT_TYPE_MISMATCH"
+            } else {
+                "FIELD_TYPE_MISMATCH"
+            };
+
+            let error = CompositionError::InternalError {
+                message: format!(
+                    "Type of {} \"{}\" is incompatible across subgraphs",
+                    element_kind,
+                    dest.coordinate()
+                ),
+            };
+
+            self.error_reporter.report_mismatch_error::<Type, ()>(
+                error,
+                &typ,
+                sources,
+                |typ, _is_supergraph| Some(format!("type \"{}\"", typ)),
+            );
+
+            Err(FederationError::SingleFederationError(
```

Comment:
I think it's sufficient to report the mismatch error to the error reporter. We can return Ok(false) to indicate that type merging failed, but not in a way we can't handle.